napari/layers/labels/labels.py

```diff
@@ -447,25 +447,28 @@ def fill(self, coord, old_label, new_label):
         ----------
         coord : sequence of float
             Position of mouse cursor in image coordinates.
-        old_label : int
-            Value of the label image at the coord to be replaced.
         new_label : int
             Value of the new label to be filled in.
         """
         self._save_history()
 
-        int_coord = np.round(coord).astype(int)
-
         if self.n_dimensional or self.ndim == 2:
             # work with entire image
             labels = self.data
-            slice_coord = tuple(int_coord)
+            slice_coord = tuple(
+                np.round(np.clip(c, 0, s - 1)).astype(int)
+                for c, s in zip(coord, self.shape)
+            )
         else:
             # work with just the sliced image
             labels = self._data_raw
-            slice_coord = tuple(int_coord[d] for d in self.dims.displayed)
+            slice_coord = tuple(
+                np.round(np.clip(coord[d], 0, self.shape[d] - 1)).astype(int)
+                for d in self.dims.displayed
+            )
 
-        matches = labels == old_label
+        self.coordinates = slice_coord
+        matches = labels == self.get_value()
```

Comment:
```suggestion
        old_label = self.get_value()
        if old_label == new_label:
            return
        matches = labels == old_label
```

See my note above, I think we should move the catch inside the fill function, lets us really make sure everything is the way we want it, and reduces the number of calls to `get_value`